addons/okr-daily-script.kp/knowledge.md

```diff
@@ -0,0 +1,222 @@
+---
+title: addon_aggregates derived dataset script
+authors:
+- Ben Miroglio
+tags:
+- add-ons
+- okr
+- derived dataset
+created_at: 2017-02-08 00:00:00
+updated_at: 2017-02-15 12:02:33.839069
+tldr: script to be run daily that contructs the addon_aggregates table in re:dash
+---
+# Add-ons 2017 OKR Data Collection
+
+Some OKRs for 2017 can be feasibly collected via the `addons` and `main_summary` tables. These tables are huge and aren't appropriate to query directly via re:dash. This script condenses these tables so that the result contains the least data possible to track the following OKRs:
+
+* **OKR 1: Increase number of users who self-install an Add-on by 5%**
+* **OKR 2: Increase average number of add-ons per profile by 3%**
+* **OKR 3: Increase number of new Firefox users who install an add-on in first 14 days by 25%**
+
+These OKRs, in addition to other add-on metrics, are tracked via the [Add-on OKRs Dashboard](https://sql.telemetry.mozilla.org/dashboard/add-on-okrs_1#edit_dashboard_dialog) in re:dash.
+
+
+```python
+from __future__ import division
+import pyspark.sql.functions as fun
+import pyspark.sql.types as st
+import pandas as pd
+import matplotlib.pyplot as plt
+import seaborn
+import math
+import datetime as dt
+
+seaborn.set_style('whitegrid')
+%matplotlib inline
+```
+
+```python
+def optimize_repartition(df, record_size, partition_size=280):
+    '''
+    Repartitions a spark DataFrame <df> so that each partition is 
+    ~ <partition_size>MB, defaulting to 280MB. record_size must be 
+    estimated beforehand--i.e. write the dataframe to s3, get the size 
+    in bytes and divide by df.count(). 
+    
+    Returns repartitioned dataframe if a repartition is necessary.
+    '''
+    total_records = df.count()
+    print "-- Found {} records".format(total_records),
+    
+    #convert megabytes to bytes
+    partition_size *= 1000000
+    
+    records_per_partition = partition_size / record_size
+    num_partitions = int(math.ceil(total_records / records_per_partition))
+
+    if num_partitions != df.rdd.getNumPartitions():
+        print "-- Repartitioning with {} partitions".format(num_partitions)
+        df = df.repartition(num_partitions)
+    return df
+
+def ping_s3(bucket, filename='result.txt'):
+    '''List the elements in <bucket> and write to file'''
+    print bucket
+    query = 'aws s3 ls --summarize --human-readable s3://'
+    os.system("{}{} | cat > {}".format(query, bucket, filename))
+
+def show_file(filename):
+    with open(filename, 'r') as f:
+        print f.read()
+        
+bool_to_int = fun.udf(lambda x: 1 if x == True else -1, st.IntegerType())
```

Comment:
I use -1 and 1 because it allows me to segment users into three groups for two different cases:

**Case 1**: Users that have only foreign-installed add-ons, only self-installed add-ons, or a combination. Summing this field grouped by `client_id` I can easily identify these groups as 1, -1, and 0 respectively by applying `boot_to_int()` on a `foreign_install` boolean.

**Case 2**: Users that have the default theme, a custom theme, or changed their theme on a given day: Using the same logic, these groups can be identified as -1, 1 and 0 respectively after applying `bool_to_int()` on a `has_custom_theme` boolean